### PR TITLE
Gtm8202

### PR DESCRIPTION
**Pull request by JakeYottaDB**

### CODE REVIEW

File: v63004/inref/gtm8202.m [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	 set ^a(4)="40"
+
+	 set ^b(3)="15"
+	 set ^b(4)="20"
+
+	 set ^c(3)="9"
+	 set ^c(4)="12"
+	 TCOMMIT
+
+	 view "flush"
```
Comment (nars1) [inline v63004/inref/gtm8202.m]: Please add a comment as to why the VIEW "FLUSH" is necessary. It is not obvious.

File: v63004/u_inref/gtm8202.csh [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+setenv gtm_test_jnl SETJNL
+setenv gtm_repl_instance mumps.repl
+setenv gtm_repl_instname INSTANCE1
+
+#setting this skips turning on replication in the passive start script
+setenv gtm_test_repl_skipsetreplic ""
+
+
+echo "#Create 3 database files for 3 regions, mumps.dat, a.dat, and b.dat for DEFAULT, AREG, and BREG respectively"
```
Comment (nars1) [inline v63004/u_inref/gtm8202.csh]: For all the "echo" statements, can we be consistent with existing protocol which is to have a # in the first column and then 1 space and then an upper-case letter followed by the rest of the line. In the above case, a space is missing between the # and C

File: v63004/u_inref/gtm8202.csh [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+$gtm_tst/com/dbcreate.csh mumps 3 255 8000 16384 1024 1024 1024 > db_log.txt
+
+$MUPIP set -repli=on  -reg "AREG" >>& db_log.txt
+$MUPIP set -repli=on  -reg "BREG" >>& db_log.txt
+$MUPIP set -repli=off -journal=disable -reg "DEFAULT" >>& db_log.txt
+
+$gtm_tst/com/passive_start_upd_enable.csh >>& passive_start.out
+
+echo "#calls gtm8202.m to set variables"
```
Comment (nars1) [inline v63004/u_inref/gtm8202.csh]: In this case, "#calls" should be "# Calls"

File: v63004/u_inref/gtm8202.csh [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+$ydb_dist/mumps -run gtm8202 > gtm8202.m.log
+
+echo "#  single region extract -SEQNO=~0,1,~2 "
```
Comment (nars1) [inline v63004/u_inref/gtm8202.csh]: In this case (and many other similar usages below), "#  single" should be "# Single"

File: v63004/u_inref/gtm8202.csh [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	rm Reg.mjf
+else
+	echo "Reg.mjf not created (not supposed to happen here)"
+endif
+
+echo "#  single region extract -SEQNO=1,2,3,4,5,6,7,8 TP transaction "
+echo "#  expecting a.lost to be created"
+$MUPIP journal -forward -extract=./Reg.mjf -seqno=1,2,3,4,5,6,7,8 "a.mjl" >>& db_log.txt;
+if (  -f "a.lost" ) then
+	echo "a.lost file created (expected)"
```
Comment (nars1) [inline v63004/u_inref/gtm8202.csh]: In addition to saying a.lost did get created, we need to display the contents of a.lost and Reg.mjf to ensure the output is the same every time around. So filter the contents of those two files just like you filter the contents in previous cases.

File: v63004/u_inref/gtm8202.csh [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+echo "#  shutdown replication and turn it back on for AREG"
+$MUPIP replic -source -shutdown -timeout=0 >>& passive_stop.out
+$gtm_tst/com/dbcheck.csh >> dbcreate_log.txt
+$MUPIP set -repli=on  -reg "AREG" >>& db_log.txt
+$MUPIP set -repli=off  -reg "BREG" >>& db_log.txt
+$MUPIP set -repli=off  -reg "DEFAULT" >>& db_log.txt
+
+echo "#  multi region extract -SEQNO=1,2,3,4 with BREG replication turned off"
+echo "#  expecting JNLEXTRCTSEQNO error"
+$MUPIP journal -forward -extract=./Reg.mjf -seqno="1,2,3,4" "*" >>& db_log.txt
```
Comment (nars1) [inline v63004/u_inref/gtm8202.csh]: The test is expecting a JNLEXTRCTSEQNO error. But the presence of that error needs to be tested here.

**Comment by nars1** [comment]
@JakeYottaDB : One more comment.

The following use case is currently not tested. Below is from the release note. Please add a test for this too (a case with unreplicated regions that works).

`For unreplicated regions, EXTRACT uses journal sequence numbers`

Thanks.

**Comment by nars1** [comment]
@JakeYottaDB : Regarding my previous comment, testing showed that in the case of unreplicated regions, the extract uses database transaction numbers (not journal sequence numbers which are not present in the case of unreplicated regions).

So the below release note should be revised

From: `For unreplicated regions, EXTRACT uses journal sequence numbers`
To : `For unreplicated regions, EXTRACT uses database transaction numbers`

Please test the revised usecase. Thanks.

File: v63004/inref/gtm8202.m [hunk 7]
```diff
@@ -31,5 +31,8 @@
 	 set ^c(4)="12"
 	 TCOMMIT
 
+	 ;ensures all updates to DB are moved from memory to disk
+	 ;mupip journal -extract only checks the disk
```
Comment (nars1) [inline v63004/inref/gtm8202.m]: mupip journal -extract --> mupip journal -extract (done in caller script right after this M program exits)"

File: v63004/outref/gtm8202.txt [hunk 8]
```diff
@@ -1,50 +1,58 @@
-#Create 3 database files for 3 regions, mumps.dat, a.dat, and b.dat for DEFAULT, AREG, and BREG respectively
-#calls gtm8202.m to set variables
-#  single region extract -SEQNO=~0,1,~2 
-#  expected to return sequences 1
-#  search extract file for set variables
+ #Create 3 database files for 3 regions, mumps.dat, a.dat, and b.dat for DEFAULT, AREG, and BREG respectively
```
Comment (nars1) [inline v63004/outref/gtm8202.txt]: Minor issue. #Create -> "# Create". Likewise in the next line #Calls.

File: v63004/u_inref/gtm8202.csh (hunk 6 above)
Comment (nars1) [reply v63004/u_inref/gtm8202.csh]: I am fine with all changes. The only one change missing is the test of unreplicated regions with -SEQNO on a single region using database transaction number.

File: v63004/u_inref/gtm8202.csh [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+else
+	echo "# Reg.mjf not created (not supposed to happen here)"
+endif
+
+echo "# Multi region extract -SEQNO=~1,2,3,4,5,6 "
+echo "# Expected to return sequences 2,3,4,5"
+$MUPIP journal -forward -extract=./Reg.mjf -seqno=~1,2,3,4,5,6 "*" >>& db_log.txt
+echo "# Search extract file for set variables"
+if (  -f Reg.mjf ) then
+	$grep "=" Reg.mjf | awk -F '\\' '{ print $6 " " $11 }'
```
Comment (nars1) [inline v63004/u_inref/gtm8202.csh]: One last standards comment. awk -> $tst_awk in various places. Sorry I did not catch this before.

File: v63004/u_inref/gtm8202.csh [hunk 10]
```diff
@@ -129,15 +132,43 @@ else
...
+else
+	echo "# Reg.mjf not created (NOT expected)"
+endif
+
+echo "# Single region extract -SEQNO="1024,1025,1026,1027,1028,1029" (DB transaction numbers) with BREG replication truned off"
+echo "# Expecting variables b(1) and b(2) to be set"
+$MUPIP journal -forward -extract=./Reg.mjf -seqno="1024,1025,1026,1027,1028,1029" "$bjnl" >>& db_log.txt
+echo "# Search extract file for set variables"
+if (  -f Reg.mjf ) then
+	$grep "=" Reg.mjf | awk -F '\\' '{ print $2 " " $11 }'
```
[the remark this answers is not in the document]
Comment (JakeYottaDB) [reply v63004/u_inref/gtm8202.csh]: The script is properly generating the expected output, I must have messed something up when updating the gtm8202.txt reference file. The issue has been corrected.

File: v63004/u_inref/gtm8202.csh [hunk 11]
```diff
@@ -129,15 +132,43 @@ else
...
+$MUPIP journal -forward -extract=./Reg.mjf -seqno="3,4" "$bjnl" >>& db_log.txt
+echo "# Search extract file for set variables"
+if (  -f Reg.mjf ) then
+	$grep "=" Reg.mjf | awk -F '\\' '{ print $6 " " $11 }'
+	rm Reg.mjf
+else
+	echo "# Reg.mjf not created (NOT expected)"
+endif
+
+echo "# Single region extract -SEQNO="1024,1025,1026,1027,1028,1029" (DB transaction numbers) with BREG replication truned off"
```
[the remark this answers is not in the document]
Comment (JakeYottaDB) [reply v63004/u_inref/gtm8202.csh]: The single DB transaction number test has been added

File: v63004/u_inref/gtm8202.csh [hunk 12]
```diff
@@ -129,15 +132,43 @@ else
...
 $MUPIP journal -forward -extract=./Reg.mjf -seqno="1,2,3,4" "*" >>& db_log.txt
 $grep "JNLEXTRCTSEQNO" db_log.txt
 
+
+echo "# Grep db_log.txt for the b.mjl_* file holding the previous transactions"
+set bjnl=`$grep 'b.mjl_' db_log.txt | awk -F '/' 'END{ print $(NF) }'`
+echo $bjnl
+
+
+echo "# Single region extract -SEQNO="3,4" (journal sequence numbers) with BREG replication truned off"
```
[the remark this answers is not in the document]
Comment (JakeYottaDB) [reply v63004/u_inref/gtm8202.csh]: The spelling errors have been corrected.

**Pull request merged**